### PR TITLE
Introduce additional -subj override, utf8 and multivalue RDN support

### PR DESCRIPTION
**Pull request by dirkx**
Replaces what was still left from PR # 1729 from 2016 (as https://github.com/openssl/openssl/commit/529586085e38487d45974817d4f3ff40f30e19f6 introduced most of this code for a slightly different usecase already).

Note to the reviewer - I am not clear why that patch also changes the Issuer around line 609. But have not changed that.

- [X] documentation is added or updated
- [X] tests are added and updated

### CODE REVIEW

**Comment by dirkx** [comment]
Sorry - this will take some time - as the tests pass on my end.

**Comment by kroeckx** [comment]
This seems related to #3531

File: apps/x509.c [hunk 1]
```diff
@@ -612,22 +624,45 @@ int x509_main(int argc, char **argv)
             goto end;
         }
 
-        n = req == NULL ? fsubj : X509_REQ_get_subject_name(req);
-        if (!X509_set_issuer_name(x, n) || !X509_set_subject_name(x, n))
+        if (req == NULL) {
+            if (fsubj == NULL) {
+                BIO_printf(bio_err, "We need either a request/certificate to extract a subject fomr; or one set with -subj\n");
```
Comment (FdaSilvaYY) [inline apps/x509.c]: Typo : fomr ?

File: apps/x509.c [hunk 2]
```diff
@@ -612,22 +624,45 @@ int x509_main(int argc, char **argv)
...
+        }
+
+        if (!X509_set_issuer_name(x, n))
             goto end;
+        if (!X509_set_subject_name(x, n))
+            goto end;
+
+        if (n == fsubj) {
+            X509_NAME_free(fsubj);
+	    fsubj = NULL;
```
Comment (FdaSilvaYY) [inline apps/x509.c]: nit : indent is wrong

**Comment by dirkx** [comment]
I'd like to pause this pull request - as I am not confident that the CI tests by AppVeyor and my own 'make test' with valgrind & electric fence are good enough. Will be adding some more test cases first.

Let me know if it is easier process wise if I close this request & re-open. Or if I should leave it here 'as is' and continue once I've gotten the right test coverage.

**Comment by richsalz** [comment]
You can close if you want.  Or edit the subject line of the PR and put "WIP:" in front, to indicate this is a work in progress.

**Comment by dirkx** [comment]
Removed the WIP - added the other testcases to confirm that the alternative (only a public key; no certificate) works correctly.

File: apps/x509.c (hunk 2 above)
Comment (dirkx) [reply apps/x509.c]: Fixed in this/last commit.

**Comment by dirkx** [comment]
Fair to assume no-one is waiting for me for further changes ?

**Comment by kroeckx** [comment]
I think we should always look at the config file as UTF-8. A command line switch should not have any effect on that. The question is just if we should enforce that it's valid UTF-8 and if we should do normalization. But I don't see how this PR has any effect on it, nor do I see any tests for it.

For command line, I think we need a general solution. I prefer one that just takes the locale settings into account and does the conversion as needed.

**Comment by dirkx** [comment]
That was how this patch originally was. 

But commit 5295860 introduced in [line 166](https://github.com/openssl/openssl/commit/529586085e38487d45974817d4f3ff40f30e19f6#diff-f48b02166abaa1faa6c10358046a8fb2R166) it set to ASC. We occasionally need to force it to UTF8. 

And I would avoid normalisation within openssl - we have situations (germany v.s. swiss springs to mind) where some entities want their normalisation subtly different from each other. So best I guess to just accept the UTF8 as is.

Happy to remove that utf8 flag from this pull request though - as needing control over UTF8 for strings that do not need it is pretty rare anyway.

**Comment by dirkx** [comment]
It is the override and multi-value RDN that we'd really like to see upstream.

**Comment by dirkx** [comment]
@kroeckx what would you prefer met to ?

- Resubmit with the utf8 flag removed; and reverting to MBSTRING_ASC (as it was prior) ?
- Resubmit with the latter set to MBSTRING_UTF8 and perhaps a flag to force ASCII (e.g. if you have a nordic or austrian issue were you want to control the normalisation to match local customs) ?

With respect to:

**Comment by kroeckx** [comment]
I suggest that you do the minimum amount of changes requires for
what you want to do, and leave all characterset related things to
some other PR.

It would also be good to hear from somebody else.

**Comment by dirkx** [comment]
Ok - will do (and found some odd effects when writing a test case for it anyway).

**Comment by dirkx** [comment]
Vacating this pull request - as someone else has introduced the UTF8 change that was to be removed from this pull request.

**Pull request closed without merging**